### ISSUE TITLE
Some questions regarding training

### ISSUE THREAD

**Issue by bernardoleite** [KristiyanVachev/Leaf-Question-Generation#1]
Hi

Thank you very much for your implementation.  It's been very helpful to me. 
Regarding your example script to generate questions and answers (google collab), I would like you to clarify some doubts for me, if possible.

1 - Do you know if it is possible to get generated tokens in ```training_step```? Instead of getting them just at the end, via ```generate()``` inference method.

2 - Do you have a particular reason to do encoding using ```tokenizer(answer + <sep> + context, ...)``` instead of using ```tokenizer(answer, context, ...)```?

3 - Have you encountered overfitting throughout your experiments? Unfortunately, my dev loss only improves up to the second epoch (cross-entropy loss of 1.35) and then increases. From what I observer from your experiments, you can reach the 4th epoch with a loss of 1.17374. 
Note:  I am using the same SQUAD v1.1 splits, model: t5-base, batch size: 32, optimizer: AdamW (eps = 1e-6). 

4 - Related to the previous question. Do you have any idea what a "good loss" is for this QG task? In any case, I haven't seen the dev loss reach a value lower than 1.

Thanks in advance.
Bernardo

**Comment by KristiyanVachev** [comment]
Hi Bernardo, thanks for your interest.

1. I don't know if that's possible with huggingface. Tough I would recommend you to check their docs, as they are really well done. For example it this [part of the tutoria](https://huggingface.co/course/chapter3/4?fw=pt)l I believe there was a way to add a function for custom evaluation, but I think it was after each epoch, not each training step.

2. Yes, I think there was a problem when I used just `tokenizer(answer, context, ...)` , because it that case the `</s>` token was added between the answer and context which messed up with how the transformer was reading it. It doesn't really matter how the separator is named, as long as you add it to the tokenizer. Also this way the transformer is learning to also put the `<sep>` in the output, which makes it easier to split the answer and question when we generate them both. 

3. Yes, this was a major problem at the start. In general what I found was that the loss did not correspond to how good the actual questions were. I did some experiments where I evaluated the BLEU score of the "best loss epoch" (e.g. 2nd epoch) and "last epoch" - (e.g.  5th epoch). In general even tough the loss was twice as bad in the last epoch, the BLEU was pretty much the same, maybe a little bit better. I didn't go any further than that, but this is an interesting topic of research in itself - which loss functions are appropriate for the task of question generation. 
In the next month I am supposed to finish my Masters thesis, so I will create some summary of my research into this and share it. I guess it's also possible for this to be picked up as a topic by some other Masters student in my university, as it seems that there is some interest in the field of question generation with us in Sofia University. Hit me up if you want me to keep you up to date :) 

4. Exactly for this reason, with my advisors we came to the conclusion that we cannot really rely on the loss alone. Especially when the model generates both answer and question it turns really weird for the loss, and it doesn't much correspond to the actual quality of the questions or the answers. 
BLEU is the most common metric for evaluating the questions, though I would still not rely much on it (another interesting research topic would be - how much does BLEU correlate with human evaluation of quality of generated questions and on translated text 😃 ), but BLEU is still a fast to compute metric and you can compare with some other major algorithms in [here](https://paperswithcode.com/sota/question-generation-on-squad11) . 
For evaluating the answers you could use Exact Match and F1 scores, but still it wouldn't be perfect as in both cases you compare to a specific versions of a question or an answer, which is pretty much okay for translation, as there aren't that much versions, but for generated questions... you can really have hundreds of good ways to ask a question.
In general and looking at other papers, seems like you don't really have train for more than 4-5 epochs on the transformer models. 

Sorry if I rambled a bit. Hit me up if you have more questions :) 

Chris

**Comment by bernardoleite** [comment]
Hey again!
Thank you very much for the quick and insightful comments.

1. It seems that we can use ```tokenizer.decode(torch.argmax(output.logits, dim=-1))``` to get the generated tokens after each epoch.  However, I still don't understand what exactly is the purpose (and effect) of passing ```output.logits``` into ```torch.argmax```. Anyway, this also allows us to use other evaluation metrics (e.g. BLEU, ROUGE and METEOR).

2. I have used ```tokenizer(answer, context, ...)``` encoding method, and apparently I had no problems or errors. What problems did you have? It seems to me that this could be a very important topic for future research. There are indeed some ways to do encoding ( for now I have seen three or four ), but how does that affect the performance of the generation models? Did you make experiments with that? Note: I am only generating the questions (not both answers and questions at the same time).

3. It looks like the cost function is really an intriguing point. Even though I get an error greater than 1, the generated questions are satisfactory (qualitative analysis). What was the best (minimum) error value you got from your experiments and at each epoch? BTW, when using Du et al. (2017) splits (and evaluation scripts) I'm getting a ```BLEU-4``` score of ```19.7```. And on your side? I will be happy to see your results in your master's thesis :) Good work!

Regards,
Bernardo

**Comment by KristiyanVachev** [comment]
Hey!

1. This is very interesting. I haven't explored this direction, but it I curious to see how you can use e.g. BLEU at each step. It would be great if you share some code/articles here for other people as well :) 

2. Indeed, there was no problem if you generate only a single output - question. But my problems arose only when I tried to generate both answer and question. But I would have to look into it further what exactly was the problem and I'm going to get back to you in the near 1 month. 

3. I found some data where I had explored exactly this difference:
![image](https://user-images.githubusercontent.com/6071722/148639301-41230664-9aac-459f-9c00-625f827e3e56.png)
This are the loss, bleu-pip (from [this package ](https://pypi.org/project/blue/)), and the Du et al. (2017) annoying python2 script run on the DEV SQuAD1.1 set. Where it is trained on the same split (where we take out the approx. 11k samples for test set). I believe this is trained on the t5-small model. But later training on t5-base got a Bleu4 score of about 15.7 after 4 epochs. 
But I never got above this score for pure question generation and moved on to other tasks such as multi-task models and distractor generation. On what t5 version (small, base, large) do you get that 19.7 score? I would be very interested if you share some details about your training/code at some point in time. You can also reach me over email : vachevkd(at)gmail.com .

 Chris

**Comment by bernardoleite** [comment]
Hello again!

Sorry for the late reply. I actually had to take a small break from this research, due to other factors.

1.  Yes, I agree with you. It is a very interesting line. However, I haven't had time to experiment with it yet, either changing the loss function to the BLEU or using it just for evaluation at each step.

3. Thanks for providing such rich information regarding your experiments. I have only experimented with Du et al. (2017) splits and evaluation scripts. Therefore, I believe that I also need to experiment with that package you have shared. It looks like there might be some discrepancies in the results when using different packages or configs.

In summary, I will be very happy to contact you when I come up with new experiences. 
Thanks again. Good work!
